### PR TITLE
[CARBONDATA-3974] Improve partition purning performance in presto carbon integration

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
a) For 200K segments table in cloud, presto partition query was taking more than 5 hours. the reason is it was reading all segment files for partition pruning. Now it is less than a minute !
b) If the query filter is < or > or more than one value. Partition pruning is not working in presto. 
c) prestodb profile compilation is broken from previous PR

 ### What changes were proposed in this PR?
a)  HiveTableHandle already have partition spec, matching for the filters (it has queried metastore to get all partitions and pruned it). So, create partitionSpec based on that. Also handled for both prestodb and prestosql  
b)  #3885 , broke prestodb compilation, only prestosql is compiled. 
c)  #3887, also didn't handled prestodb

    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No [Need to add spark support and create better UT for presto, TODO]
verified manually

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3979/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2239/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3984/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2244/

**Comment by ajantha-bhat** [comment]
retest this please

**Comment by ajantha-bhat** [comment]
PR is ready. Please review @QiangCai , @kunal642

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3990/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2251/

**Comment by marchpure** [comment]
retest this please

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java [hunk 1]
```diff
@@ -117,6 +122,16 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHand
       // file metastore case tablePath can be null, so get from location
       location = table.getStorage().getLocation();
     }
+    List<PartitionSpec> filteredPartitions = new ArrayList<>();
```
Comment (Indhumathi27) [inline integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:125]: Can you add a testcase with partition filter?

Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:125]: please read the description, I have mentioned why UT cannot be added now

File: integration/presto/src/main/prestosql/org/apache/carbondata/presto/impl/CarbonTableReader.java [hunk 2]
```diff
@@ -245,16 +242,14 @@ private CarbonTableCacheModel getValidCacheBySchemaTableName(SchemaTableName sch
...
    * @return list of multiblock split
    * @throws IOException
    */
-  public List<CarbonLocalMultiBlockSplit> getInputSplits(
-      CarbonTableCacheModel tableCacheModel,
-      Expression filters,
-      TupleDomain<HiveColumnHandle> constraints,
-      Configuration config) throws IOException {
+  public List<CarbonLocalMultiBlockSplit> getInputSplits(CarbonTableCacheModel tableCacheModel,
+      Expression filters, List<PartitionSpec> filteredPartitions, Configuration config)
```
Comment (Indhumathi27) [inline integration/presto/src/main/prestosql/org/apache/carbondata/presto/impl/CarbonTableReader.java:251]: Can revert to old style

File: integration/presto/src/test/prestodb/org/apache/carbondata/presto/server/PrestoTestUtil.scala [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.presto.server
+
+import com.facebook.presto.jdbc.PrestoArray
+
```
Comment (Indhumathi27) [inline integration/presto/src/test/prestodb/org/apache/carbondata/presto/server/PrestoTestUtil.scala:20]: Remove extra lines

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4003/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2263/

**Comment by marchpure** [comment]
I just tested. With this PR. Query nonpartition table will has EMPTY RESULT. Query  parititon table works well

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataModule.java [hunk 4]
```diff
@@ -21,6 +21,8 @@
 
 import static java.util.Objects.requireNonNull;
 
+import org.apache.carbondata.core.constants.CarbonCommonConstants;
+import org.apache.carbondata.core.util.CarbonProperties;
```
Comment (kunal642) [inline integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataModule.java:25]: Why this change is required?

Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataModule.java:25]: prestodb compile was broken by #3885 , They have copied code but not import statement. so, fixed it. Mentioned in PR description also

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java [hunk 5]
```diff
@@ -117,6 +122,16 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transactionHand
...
+    if (layout.getPartitionColumns().size() > 0 && layout.getPartitions().isPresent()) {
+      List<String> colNames =
+          layout.getPartitionColumns().stream().map(x -> ((HiveColumnHandle) x).getName())
+              .collect(Collectors.toList());
+      for (HivePartition partition : layout.getPartitions().get()) {
+        filteredPartitions.add(new PartitionSpec(colNames,
```
Comment (kunal642) [inline integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]: according to this code all the PartitionSpec will have all the column names!!! is this correct?
The column names in partitionSpec should be per partition right?

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+import com.facebook.presto.spi.block.BlockBuilder;
+
+import org.apache.carbondata.presto.CarbonVectorBatch;
+import org.apache.carbondata.presto.ColumnarVectorWrapperDirect;
+
+/**
+ * Class to read the complex type Stream [array/struct/map]
+ */
+
+public class ComplexTypeStreamReader extends CarbonColumnVectorImpl
```
Comment (kunal642) [inline integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java:45]: This change is too big and not related to partition pruning, please raise a seperate PR for this.

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java (hunk 5 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]: you mean all the partition column right ? not the all table columns ?

Ideally they use partitionPath. I have tested it. It didn't impacted. Let me check again.

Comment (kunal642) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]: yeah all partition column, 1 partition spec should have only the columns it is created on, not all columns.

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java (hunk 6 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java:45]: Then I have to raise partition changes only for prestosql, because prestodb profile is not compiling.  or I fix the prestodb compile first. you can merge it. I will raise separate PR for partition. what you suggest ?

Comment (kunal642) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java:45]: is this compilation fix? if not then keep the compilation changes in this PR and ComplexTypeStreamReader looks to be a bug fix!!!

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java (hunk 5 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]: I think HiveTableHandle will give matching column for the partitions only , because this is per query derived.

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java (hunk 6 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java:45]: Not a bug fix, you need to know how prestodb and prestosql is divided.
there are some common files and individual files.  This class is individual class

There is a complexStreamReader in prestoSql also. same code but import packages are from prestosql. import packages are from prestodb

Comment (kunal642) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/readers/ComplexTypeStreamReader.java:45]: ok

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataModule.java (hunk 4 above)
Comment (kunal642) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataModule.java:25]: ok

File: integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java (hunk 5 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]: I have verified. This scenario can work. I guess. No need to modify any code.

Comment (ajantha-bhat) [reply integration/presto/src/main/prestodb/org/apache/carbondata/presto/CarbondataSplitManager.java:131]:
```
presto:redods> select dtm,hh from dw_log_ubt_partition_carbon_neww;
     dtm     |     hh
-------------+------------
 part_dtm_01 | part_hh_01
 part_dtm_01 | part_hh_01
 part_dtm_01 | part_hh_02
 part_dtm_20 | part_hh_21
 part_dtm_01 | part_hh_03
 part_dtm_21 | NULL
(6 rows)

Query 20200910_035416_00017_wv9qh, FINISHED, 3 nodes
Splits: 22 total, 22 done (100.00%)
0:01 [6 rows, 176B] [9 rows/s, 282B/s]

presto:redods> select dtm,hh from dw_log_ubt_partition_carbon_neww where (dtm = 'part_dtm_01' and hh = 'part_hh_03') or dtm='part_dtm_21';
     dtm     |     hh
-------------+------------
 part_dtm_01 | part_hh_03
 part_dtm_21 | NULL
(2 rows)

Query 20200910_035548_00018_wv9qh, FINISHED, 3 nodes
Splits: 18 total, 18 done (100.00%)
0:01 [2 rows, 0B] [1 rows/s, 0B/s]


```

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4038/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2299/

File: integration/presto/src/main/prestosql/org/apache/carbondata/presto/impl/CarbonTableReader.java (hunk 2 above)
Comment (ajantha-bhat) [reply integration/presto/src/main/prestosql/org/apache/carbondata/presto/impl/CarbonTableReader.java]: done

File: integration/presto/src/test/prestodb/org/apache/carbondata/presto/server/PrestoTestUtil.scala (hunk 3 above)
Comment (ajantha-bhat) [reply integration/presto/src/test/prestodb/org/apache/carbondata/presto/server/PrestoTestUtil.scala:20]: done

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4051/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2313/

**Comment by kunal642** [comment]
LGTM
